src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs

```diff
@@ -31,63 +73,153 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
                 Name = "Console logger queue processing thread"
             };
             _outputThread.Start();
+            _processorId = ++s_processorCount;
+            s_processors.Add(this);
         }
 
-        public virtual void EnqueueMessage(LogMessageEntry message)
+        static ConsoleLoggerProcessor()
         {
-            if (!_messageQueue.IsAddingCompleted)
-            {
-                try
-                {
-                    _messageQueue.Add(message);
-                    return;
-                }
-                catch (InvalidOperationException) { }
-            }
+            s_meter.CreateObservableGauge<long>("queue-size", GetQueueSize);
+        }
 
-            // Adding is completed so just log the message
-            try
+        public virtual void EnqueueMessage(LogMessageEntry message)
+        {
+            // cannot enqueue when adding is completed
+            if (!Enqueue(message))
             {
                 WriteMessage(message);
             }
-            catch (Exception) { }
         }
 
         // for testing
         internal void WriteMessage(LogMessageEntry entry)
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            try
+            {
+                IConsole console = entry.LogAsError ? ErrorConsole : Console;
+                console.Write(entry.Message);
+            }
+            catch (Exception)
+            {
+                CompleteAdding();
+            }
         }
 
         private void ProcessLogQueue()
         {
-            try
+            while (!_isAddingCompleted || _messageQueue.Count > 0)
```

Comment:
Won't the thread complete if the `_messageQueue` is empty?

Reply:
because that immediately takes us out before the first log message gets hit. that is not what we want